cql3/statements/describe_statement.cc

```diff
@@ -160,14 +161,78 @@ description type(replica::database& db, const lw_shared_ptr<keyspace_metadata>&
     return description(db, *udt, true);
 }
 
-future<std::vector<description>> types(replica::database& db, const lw_shared_ptr<keyspace_metadata>& ks, bool with_stmt = false) {
-    auto user_types_map = ks->user_types().get_all_types();
-    auto user_types = boost::copy_range<std::vector<shared_ptr<const keyspace_element>>>(user_types_map | boost::adaptors::transformed([] (const auto& e) {
-        // return static_pointer_cast<const keyspace_element>(e.second);
-        return e.second;
-    }));
+/*
+ * Sort user-defined types topologically.
+ * If a UDT `a` is used in any field in a UDT `b`,
+ * then `a` should be before `b`.
+ *
+ * There shouldn't be any cycles as we don't allow any
+ * circular dependencies in UDTs.
+ *
+ * Algorithm:
+ * - for each UDT, count in how many other UDTs it's used
+ * - push to result UDTs which counter is 0 (they aren't used in any other UDTs)
+ * - for each UDT `u` in current result vector:
+ *   - get UDTs `ref_udt`s used in `u`
+ *   - decrement counter of `ref_udt`s
+ *   - if counter of any `ref_udt` drops to 0, push that `ref_udt` to end of result vector
```

Comment:
Isn't there an outer loop?

Reply:
No, new UDTs are added to the end of result vector as I iterate over the vector. If the are no cycles, all UDTs will be visited.